src/kubernetes/reflector.rs

```diff
@@ -24,26 +24,53 @@ pub async fn custom_reflector<K, W>(
     let mut delay_queue = DelayQueue::default();
     loop {
         tokio::select! {
-            Some(Ok(event)) = stream.next() => {
-                match event {
-                    watcher::Event::Applied(_) => {
-                        // Immediately apply event
-                        store.apply_watcher_event(&event);
-                    }
-                    watcher::Event::Deleted(_) => {
-                        // Delay reconciling any `Deleted` events
-                        delay_queue.insert(event.to_owned(), delay_deletion);
-                    }
-                    watcher::Event::Restarted(_) => {
-                        // Clear all delayed events when the cache is refreshed
-                        delay_queue.clear();
-                        store.apply_watcher_event(&event);
-                    }
+            result = stream.next() => {
+                match result {
+                    Some(Ok(event)) => {
+                        match event {
+                            // Immediately reoncile `Applied` event
+                            watcher::Event::Applied(_) => {
+                                trace!("Processing Applied event.");
+                                store.apply_watcher_event(&event);
+                            }
+                            // Delay reconciling any `Deleted` events
+                            watcher::Event::Deleted(_) => {
+                                trace!("Queuing Deleted event.");
+                                delay_queue.insert(event.to_owned(), delay_deletion);
+                            }
+                            // Clear all delayed events on `Restarted` events
+                            watcher::Event::Restarted(_) => {
+                                trace!("Processing Restarted event.");
+                                delay_queue.clear();
+                                store.apply_watcher_event(&event);
+                            }
+                        }
+                    },
+                    Some(Err(error)) => {
+                        warn!(message = "Watcher stream got an error.", error = ?error);
+                    },
+                    None => {
+                        info!("Watcher stream has terminated.");
+                        break;
+                    },
                 }
             }
-            // When the Deleted event expires from the queue pass it to the store
-            Some(Ok(event)) = delay_queue.next() => {
-                store.apply_watcher_event(&event.into_inner())
+            result = delay_queue.next(), if !delay_queue.is_empty() => {
```

Comment:
I would drop this conditional otherwise it makes the `select` arms in-exhaustive which is what led to the bug this is fixing. Instead have the `None` arm below just do nothing.

Reply:
I don't think the arms are non-exhaustive with this precondition, as I still included the `None` outcome that happens if the queue is empty - just logged that we shouldn't have reached that outcome.

However, this was @lukesteensen's suggestion on not just constantly polling the delay queue when it's empty.

Reply:
Yeah, thinking through this more, it should be ok. If the queue is empty, this arm will be disabled, and Tokio will wait on the other arm, i.e. for an event from the watcher.

It does make the below `None` arm unreachable and so, again, I might use `unreachable` there 😄